library/ui/src/main/java/com/google/android/exoplayer2/ui/PlayerControlView.java

```diff
@@ -832,21 +836,39 @@ private void updateProgress() {
       long delayMs;
       if (player.getPlayWhenReady() && playbackState == Player.STATE_READY) {
         float playbackSpeed = player.getPlaybackParameters().speed;
-        if (playbackSpeed <= 0.1f) {
-          delayMs = 1000;
-        } else if (playbackSpeed <= 5f) {
-          long mediaTimeUpdatePeriodMs = 1000 / Math.max(1, Math.round(1 / playbackSpeed));
+        int timeBarWidth = timeBar.getTimeBarWidthDp();
+
+        if (timeBarWidth == 0 || duration == 0) {
+          delayMs = MAX_UPDATE_FREQUENCY_MS;
+        } else {
+          // Calculate how many updates needs to be done with DEFAULT_UPDATE_DP
+          // to fill up the time bar
+          int numberOfUpdates = timeBarWidth / DEFAULT_UPDATE_DP;
+
+          // Calculate the designated update interval, taking duration into consideration as well
+          long mediaTimeUpdatePeriodMs = duration / numberOfUpdates;
```

Comment:
Sorry for being nitty about this, but this can still result in divisions by zero.  For example if timeBarWidth==3. The same is true for the other two divisions on line 852 and 861. 

Maybe the easiest solution is to move this code to a separate private static method getTimeBarUpdateDelayMs. Then you do the steps and return early if an intermediate value is outside the valid range. This avoids a lot of nested if conditions. Something like:
int numberOfUpdate = ...
if (numberOfUpdates <= 0) {
  return MAX_UPDATE_FREQUENCY_MS;
}
etc.

[ We usually put static methods below non-static ones, so please move that new method down to the other static methods. Thanks. ]